NEWS.md

```diff
@@ -0,0 +1,159 @@
+  processing have improved.
+* Handling of half-closed streams has changed (READ THIS SECTION).
+* Some constructors and method signatures have changed slightly.
+
+#### Bandwidth Metrics
+
+To start out on an unhappy note, bandwidth metrics are now less accurate. In the
+past, transports returned "minimal" connections (e.g., a TCP connection) so we
+could wrap these transport connections in "metrics" connections that counted
+every byte sent and received.
+
+Unfortunately, now that we've moved encryption and multiplexing down into the
+transport layer, the connection we're wrapping has significantly more
+under-the-covers overhead.
+
+However, we do hope to improve this and get even *better* bandwidth metrics than
+we did before. See
+[libp2p/go-libp2p-transport#31](https://github.com/libp2p/go-libp2p-transport/issues/31)
+for details.
+
+#### Notifications
+
+This release brings new, performant, easy to understand guarantees around libp2p
+connect/disconnect event ordering:
+
+1. For any given connection/stream, libp2p will wait for all connect/open event
+   handlers to finish exit before triggering a disconnect/close event for the
+   connection/stream.
+2. When a user calls the `Close` (or `Reset`) method on a connection or stream,
+   go-libp2p will process the close event asynchronously (i.e., not block the
+   call to `Close`). Otherwise, a call to `Close` from within a connect event
+   handler would deadlock.
+3. Unless otherwise noted, events will be handled in parallel.
+
+What does this mean for end users? Well:
+
+1. Reference counting connections to a peer using connect/disconnect events
+   should "just work" and should never go negative.
+2. Under heavy connect/disconnect loads, connecting to new peers should be
+   faster (usually).
+
+In the past, (dis)connect and stream open/close notifications have been a bit of
+a pain point. For a long time, they were fired of in parallel and one could, for
+example, process a disconnect notification before a connect notification (we had
+to support *negative* ref-counts in several places to account for this).
+
+After no end of trouble, we finally "fixed" this by synchronizing notification
+delivery. We still delivered notifications to all notifiees in parallel, we just
+processed the events in series.
+
+Unfortunately, under heavy connect/disconnect load, new connections could easily
+get stuck on open behind a queue of connect events all being handled in series.
+In theory, these events should have been handled quickly but in practice, it's
+very hard to avoid locks *entirely* (bitswap's event handlers were especially
+problematic).
+
+Worse, this serial delivery guarantee didn't actually provide us with an
+*in-order* delivery guarantee as it was still possible for a disconnect to
+happen before we even *started* to fire the connect event. The situation was
+slightly better than before because the events couldn't overlap but still far
+from optimal.
```

Comment:
maybe try to end this section on a happier note. "This has all been resolved... and should be super good now"

Reply:
Yeah, the entire thing could use a bit more "pep" (I was tired).